src/utils/graph/constraints.js

```diff
@@ -11,96 +11,145 @@ import { distance1d, greaterOrEqual, equalTo, subtract } from './common';
  */
 export const rowConstraint = {
   property: 'y',
-  difference: subtract,
-  distance: distance1d,
-  operator: greaterOrEqual,
-  target: (a, b, co, constants) => constants.spaceY,
-  strength: () => 1,
-  weightA: () => 0,
-  weightB: () => 1,
-  required: true
+
+  solve: (constraint, constants) => {
+    const { a, b } = constraint;
+    const difference = a.y - b.y;
+    const target = constants.spaceY;
+
+    if (difference >= target) {
+      return;
+    }
+
+    const resolve = difference - target;
+    a.y -= 0.5 * resolve;
+    b.y += 0.5 * resolve;
+  },
+
+  strict: (constraint, constants, variableA, variableB) => {
+    return new Constraint(
+      variableA.minus(variableB),
+      Operator.Ge,
+      constants.spaceY,
+      Strength.required
+    );
+  }
 };
 
 /**
  * Layout constraint in Y for separating layers
  */
 export const layerConstraint = {
   property: 'y',
-  difference: subtract,
-  distance: distance1d,
-  operator: greaterOrEqual,
-  target: (a, b, co, constants) => constants.layerSpace,
-  strength: () => 1,
-  weightA: () => 0,
-  weightB: () => 1,
-  required: false
+
+  solve: (constraint, constants) => {
+    const { a, b } = constraint;
+    const difference = a.y - b.y;
+    const target = constants.layerSpace;
```

Comment:
This `solve` method looks very very similar to the previous one - is it worth making a factory function for it and passing it an argument to create the resulting method? e.g. something like:

```es6
const getSolve = (targetKey) => {
  return (constraint, constants) => {
    const { a, b } = constraint;
    const difference = a.y - b.y;
    const target = constants[targetKey];

    if (difference >= target) {
      return;
    }

    const resolve = difference - target;
    a.y -= 0.5 * resolve;
    b.y += 0.5 * resolve;
  }
};


export const rowConstraint = {
  solve: getSolve('spaceY')
  ...
};


export const layerConstraint = {
  solve: getSolve('layerSpace')
  ...
};
```

If it's too much abstraction, fair enough.

Reply:
In this case I think I'm aiming to avoid too much abstraction in favour of flexibility and clarity, but once the API solidifies a bit more I'll see if it makes sense to find a natural abstraction for the resulting approach.

Reply:
Fair enough, I thought you might say that.